x/ccv/provider/keeper/relay.go

```diff
@@ -213,57 +222,65 @@ func (k Keeper) EndBlockCIS(ctx sdk.Context) {
 	k.SetValsetUpdateBlockHeight(ctx, valUpdateID, uint64(ctx.BlockHeight()+1))
 }
 
-// OnRecvSlashPacket slashes and jails the given validator in the packet data
+// OnRecvSlashPacket receives a slash packet, validates it, then handles it if valid.
 func (k Keeper) OnRecvSlashPacket(ctx sdk.Context, packet channeltypes.Packet, data ccv.SlashPacketData) exported.Acknowledgement {
-	// check that the channel is established
-	chainID, found := k.GetChannelToChain(ctx, packet.DestinationChannel)
-	if !found {
-		// SlashPacket packet was sent on a channel different than any of the established CCV channels;
-		// this should never happen
-		panic(fmt.Errorf("SlashPacket received on unknown channel %s", packet.DestinationChannel))
+
+	if err := k.ValidateSlashPacket(ctx, packet, data); err != nil {
+		return channeltypes.NewErrorAcknowledgement(err.Error())
 	}
 
 	// apply slashing
-	if _, err := k.HandleSlashPacket(ctx, chainID, data); err != nil {
-		errAck := channeltypes.NewErrorAcknowledgement(err.Error())
-		return &errAck
-	}
+	chainID := k.getChainIdOrPanic(ctx, packet)
+	k.HandleSlashPacket(ctx, chainID, data)
 
-	ack := channeltypes.NewResultAcknowledgement([]byte{byte(1)})
-	return ack
+	return channeltypes.NewResultAcknowledgement([]byte{byte(1)})
 }
 
-// HandleSlashPacket slash and jail a misbehaving validator according the infraction type
-func (k Keeper) HandleSlashPacket(ctx sdk.Context, chainID string, data ccv.SlashPacketData) (success bool, err error) {
-	// map VSC ID to infraction height for the given chain ID
-	var infractionHeight uint64
-	var found bool
-	if data.ValsetUpdateId == 0 {
-		infractionHeight, found = k.GetInitChainHeight(ctx, chainID)
-	} else {
-		infractionHeight, found = k.GetValsetUpdateBlockHeight(ctx, data.ValsetUpdateId)
-	}
+// validateSlashPacket validates a recv slash packet before it is
+// handled or persisted in store. An error is returned if the packet is invalid,
+// and an error ack should be relayed to the sender.
+func (k Keeper) ValidateSlashPacket(ctx sdk.Context,
```

Comment:
This is where the new check will be added to address https://github.com/cosmos/interchain-security/issues/546